sbin/dset/dset.8
.\"
.\" Copyright (c) 1996 Joerg Wunsch
.\"
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE DEVELOPERS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE DEVELOPERS BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" $Id: dset.8,v 1.7 1998/06/03 04:21:39 jkoshy Exp $
.\" "
.Dd May 2, 1996
.Dt DSET 8
.Os FreeBSD
.Sh NAME
.Nm dset
.Nd "update the boot file to the current configuration"
.Sh SYNOPSIS
.Nm dset
.Op Fl vqt
.Sh DESCRIPTION
.Nm Dset
records the configuration changes for ISA devices that have been made
from
.Em UserConfig
after booting with the
.Fl c
option.  The changes (if any) are recorded to the file name returned by
.Xr getbootfile 3 .
.Nm Dset
is typically called from within
.Pa /etc/rc .
.Pp
.Nm Dset
understands the following options:
.Bl -tag -width indent -compact
.It Fl v
Verbose; tell what is being done.
.It Fl q
Quiet; don't print any error messages.
.It Fl t
Test only; don't actually record any changes to the boot file, print
only what would have been done.  This implies
.Fl v .
.El
.Sh DIAGNOSTICS
The
.Nm
utility exits with status 0 on success, or with status 1 if any
problem occurred.  Unless
.Fl q
has been specified, a short message hinting to the cause of the problem
will be printed to standard error output.
.Sh SEE ALSO
.Xr getbootfile 3 ,
.Xr boot_i386 8 ,
.Xr rc 8
.Sh HISTORY
The
.Nm
command appeared in
.Fx 2.0.5 .


NEWS
# Major changes between releases

## Changes in version 0.15.0

**Released on 2026/10/07.**

### Bugfixes

* #265: Null cleanups
* #271: kyua: Try harder to delete directories.
* #300: Implement utils::config::positive_int_node::deep_copy

### Developer Workflow Changes

* #254: Add initial GitHub Actions support for kyua
* #288: [CI] remove legacy CI scripts and files
* #291: Remove Travis CI config
* #298: Add sanitizers and code coverage support
* #301: Add FreeBSD CI support via vmactions
* #305: Automate uploading release assets

### Feature Enhancements

* #203: Add GoogleTest support
* #230: Do not count skipped as passed in test cmd
* #260: Support require.diskspace ATF metadata
* #261: Pass unprivileged user config prop to ATF using all known names
* #266: Add "kyua debug -p" option
* #270: Add required_kmods metadata
* #272: Make "test" command actually report about workdir cleanup issues
* #273: Make "debug -p" skip writing stdout/stderr to tmp files
* #277: kyua-debug: Add -P option
* #290: Require C++-20 in order to build kyua
* #307: Upstream multiple items from FreeBSD

### Documentation

* #258: kyuafile.5: Add ATF metadata mapping reference
* #275: kyua manuals: Correct --results-file short form
* #279: docs: Standardize document description tone

### Miscellaneous

* #264: Switch from std::auto_ptr<> to std::unique_ptr<>
* #269: Fix cmd_about_test:topic__version test case
* #297: Reduce overall lutok dependency
* #312: Replace KYUA_ATTRIBUTE_{NORETURN,UNUSED}} with C++ attributes
* #318: utils::fs: remove umount(8) support

**Full Changelog**: https://github.com/freebsd/lutok/compare/kyua-0.14.1...kyua-0.15.0

## Changes in version 0.14.1

**Released on March 29th, 2025.**

* Issue #263: fix release tarball by relying on `make dist`.

* Document ATF to Kyua mappings for metadata properties.

* Support `require.diskspace` ATF metadata property.

* Support backwards compatible `unprivileged_user` metadata property for ease
  of use dealing with ATF vs Kyua.

* Fix building from source by tracking `utils/process/Kyuafile.in` and
  `engine/execenv/execenv_fwd.hpp` with automake.

## Changes in version 0.14

**Released on December 31st, 2024.**

* Replace `--with-atf` with `--enable-atf`. The latter must be specified
  to enable ATF support and the Kyua tests by proxy.

* Require newer versions of ATF (0.21) and Lutok (0.5). This is being done
  as part of a multi-phase deprecation of pre-C++11 environments.

* Vendor the ATF m4 files from the freebsd/atf repo.

* Translate `is_exclusive` metadata to its ATF counterpart.

* Issue #236: fix accidental test forkbomb in specific user environments
  with `integration` tests by avoiding infinite recursion calling kyua.

* Fix the build on Ubuntu 24.04.

- execenv support added. This FreeBSD-specific feature allows tests to be
  executed in jails.

* Issue #169: Remove deprecated `readdir_r` use.

* Fix build issue on more recent versions of MacOS.

* Fix build issue on IllumOS.

* Fix issues with PID reuse in testcases.

* Escape characters captured from stderr and stdout in HTML output.

* `integration/cmd_report_test:verbose`: fix regular expression used in
  tests which would break in certain user environments.

* Add Cirrus CI support for testing Kyua on FreeBSD.

* Remove `UTILS_UNUSED_PARAM` macro.

* Fix document rendering with newer versions of doxygen.

* Issue #152: explicitly require C++11 language features when compiling Kyua.
  Remove `std::shared_ptr` compatibility header, `utils/shared_ptr.hpp`.

* Fix `igor` and `mandoc -Tlint` reported issues with generated manpages.

* Issue #167: fix typos in kyua(1) manpage.

## Changes in version 0.13

**Released on August 26th, 2016.**

* Fixed execution of test cases as an unprivileged user, at least under
  NetBSD 7.0.  Kyua-level failures were probably a regression introduced
  in Kyua 0.12, but the underlying may have existed for much longer:
  test cases might have previously failed for mysterious reasons when
  running under an unprivileged user.

* Issue #134: Fixed metadata test broken on 32-bit platforms.

* Issue #139: Added per-test case start/end timestamps to all reports.

* Issue #156: Fixed crashes due to the invalid handling of cleanup
  routine data and triggered by the reuse of PIDs in long-running Kyua
  instances.

* Issue #159: Fixed TAP parser to ignore case while matching `TODO` and
  `SKIP` directives, and to also recognize `Skipped`.

* Fixed potential crash due to a race condition in the unprogramming of
  timers to control test deadlines.


## Changes in version 0.12

**Released on November 22nd, 2015.**

This is a huge release and marks a major milestone for Kyua as it finally
implements a long-standing feature request: the ability to execute test
cases in parallel.  This is a big deal because test cases are rarely
CPU-bound: running them in parallel yields much faster execution times for
large test suites, allowing faster iteration of changes during development.

As an example: the FreeBSD test suite as of this date contains 3285 test
cases.  With sequential execution, a full test suite run takes around 12
minutes to complete, whereas on a 4-core machine with a high level of
parallelism it takes a little over 1 minute.

Implementing parallel execution required rewriting most of Kyua's core and
partly explains explains why there has not been a new release for over a
year.  The current implementation is purely subprocess-based, which works
but has some limitations and has resulted in a core that is really complex
and difficult to understand.  Future versions will investigate the use of
threads instead for a simplified programming model and additional
parallelization possibilities.

* Issue #2: Implemented support to execute test cases in parallel when
  invoking `kyua test`.  Parallel execution is *only* enabled when the new
  `parallelism` configuration variable is set to a value greater than `1`.

sbin/dset/dset.c
/*
 * Copyright (c) 1995 Ugen J.S.Antsilevich
 *
 * Redistribution and use in source forms, with and without modification,
 * are permitted provided that this entire comment appears intact.
 *
 * Redistribution in binary form may occur without any restrictions.
 * Obviously, it would be nice if you gave credit where credit is due
 * but requiring it would be too onerous.
 *
 * This software is provided ``AS IS'' without any warranties of any kind.
 *
 * Device configuration to kernel image saving utility.
 */

#ifndef lint
static const char rcsid[] =
	"$Id: dset.c,v 1.14 1998/10/06 07:32:45 msmith Exp $";
#endif /* not lint */

#include <err.h>
#include <fcntl.h>
#include <kvm.h>
#include <a.out.h>
#include <limits.h>
#include <paths.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/param.h>
#include "i386/isa/isa_device.h"

#include "i386/isa/pnp.h"
struct pnp_cinfo new_ov[MAX_PNP_LDN];
struct pnp_cinfo old_ov[MAX_PNP_LDN];

#define TRUE	1
#define FALSE 	0

struct nlist    nl[] = {
#define N_TABTTY 	0
	{"_isa_devtab_tty"},
#define N_TABBIO	1
	{"_isa_devtab_bio"},
#define N_TABNET	2
	{"_isa_devtab_net"},
#define N_TABCAM	3
	{"_isa_devtab_cam"},
#define N_TABNULL	4
	{"_isa_devtab_null"},
	{""},
};
#define N_TABLAST	N_TABNULL

struct nlist    nlk[] = {
	{"_isa_devlist"},
	{""},
};

struct nlist	nlaux[] = {
	{"_num_eisa_slots"},
	{""},
};

struct nlist	nlpnp[] = {
	{"_pnp_ldn_overrides"},
	{""},
};

struct nlist	nlconfig[] = {
	{"config.o"},
	{""},
};

int             quiet = FALSE;

void
fatal(name, str)
	char           *name, *str;
{
	if (quiet)
		exit(1);
	if (str)
		errx(1, "%s: %s", name, str);
	else
		errx(1, "%s", name);
}

void
error(name, str)
	char           *name, *str;
{
	if (quiet)
		return;
	if (str)
		warnx("%s: %s", name, str);
	else
		warnx("%s", name);
}

void
usage()
{
	fprintf(stderr, "usage: dset [-qtv]\n");
	exit(1);
}

int
main(ac, av)
	int             ac;
	char          **av;
{
	int             f, res, s, i;
	int             modified,dev_found;
	int             sym;
	u_long          pos, entry, pos1, pos_t;
	u_long          flags;
	struct isa_device buf, buf1;
	struct isa_driver dbuf;
	char            nbuf[5];
	struct stat     fst;
	struct exec     es;
	kvm_t          *kd;
	static char     errb[_POSIX2_LINE_MAX];
	const char     *kernel = NULL;

	char            ch;
	int             testonly = FALSE;
	int             verbose = FALSE;

	while ((ch = getopt(ac, av, "qtv")) != -1)
		switch (ch) {
		case 'q':
			quiet = TRUE;
			break;
		case 't':
			testonly = TRUE;
			/* In test mode we want to be verbose */

		case 'v':
			verbose = TRUE;
			break;

		case '?':
		default:
			usage();
		}


	kernel = getbootfile();
	if (verbose)
		printf("Boot image: %s\n", kernel);

	if (!(kd = kvm_open(NULL, NULL, NULL, O_RDONLY, errb)))
		fatal("kvm_open", NULL);

	if (kvm_nlist(kd, nlk) != 0)
		fatal("kvm_nlist", NULL);

	if (nlk[0].n_type == 0)
		fatal("kvm_nlist", "bad symbol type");

	if (nlist(kernel, nl) != 0)
		fatal("nlist", NULL);

	if (nl[0].n_type == 0)
		fatal("nlist", "bad symbol type");

	if (stat(kernel, &fst) < 0)
		fatal("stat", NULL);

	flags = fst.st_flags;

	if (chflags(kernel, (u_long) 0) < 0)
		fatal("chflags", NULL);

	if ((f = open(kernel, O_RDWR)) <= 0)
		fatal("open", NULL);

	if (read(f, &es, sizeof(struct exec)) <= 0)
		fatal("read header", NULL);

	entry = es.a_entry;

	for (sym = 0; sym <= N_TABLAST; sym++) {
		if (verbose)
			printf("\nTable: %s\n", nl[sym].n_name);
		pos = nl[sym].n_value + getpagesize() - entry;

		pos1 = nlk[0].n_value;

		if (lseek(f, pos, SEEK_SET) != pos)
			fatal("seek", NULL);

		if (verbose)
			printf("----------------------------------------------------\n");

		do {
			if ((res = read(f, (char *) &buf, sizeof(struct isa_device)))
			    <= 0)
				fatal("read", NULL);



		if (kvm_read(kd, pos1, &pos_t, sizeof(u_long)) < 0)
			fatal("kvmread", NULL);
		dev_found = 0;

		while(pos_t!=0) {
			if (kvm_read(kd, pos_t, &buf1, sizeof(struct isa_device)) < 0)
				fatal("kvmread", NULL);

			if (buf1.id_id != buf.id_id) {
				pos_t = (u_long)(buf1.id_next);
				continue;
			} else
				dev_found=1;

			if (buf1.id_driver)
				if (kvm_read(kd, (u_long) buf1.id_driver,
					     &dbuf, sizeof(struct isa_driver)) < 0) {
					error("kvm_read", "no driver");
				} else {
					if (kvm_read(kd, (u_long) dbuf.name,
					  nbuf, sizeof(nbuf)) < 0) {
						error("kvm_read", NULL);
					} else {
						nbuf[sizeof(nbuf) - 1] = 0;
						if (verbose)
							printf("Device: %s%d\n", nbuf, buf1.id_unit);
					}
				}
			else
				error("kvm_read", "no driver");
			break;

		};

		if (!dev_found)
			continue;

			if (buf1.id_id != 0)
				if (verbose)
					printf(
  "kernel: id=%u io=%X irq=%d drq=%d maddr=%p msize=%d flags=%X enabled=%X \n",
	buf1.id_id, buf1.id_iobase, buf1.id_irq, buf1.id_drq,
	buf1.id_maddr, buf1.id_msize, buf1.id_flags, buf1.id_enabled);

			if (buf.id_id != 0)
				if (verbose)
					printf(
  "file: id=%u io=%X irq=%d drq=%d maddr=%p msize=%d flags=%X enabled=%X \n",
	buf.id_id, buf.id_iobase, buf.id_irq, buf.id_drq,
	buf.id_maddr, buf.id_msize, buf.id_flags, buf.id_enabled);


			/*
			 * OK,now we'd compare values and set'em from kernel.
			 */
			modified = FALSE;

			if (buf.id_iobase != -1 && buf.id_iobase !=
			    buf1.id_iobase) {
				if (verbose)
					printf("Setting IO addr\n");
				buf.id_iobase = buf1.id_iobase;
				modified = TRUE;
			}
			if (buf.id_irq != (u_short)-1 && buf.id_irq != buf1.id_irq) {
				if (verbose)
					printf("Setting IRQ\n");
				buf.id_irq = buf1.id_irq;
				modified = TRUE;
			}
			if (buf.id_drq != -1 && buf.id_drq != buf1.id_drq) {
				if (verbose)
					printf("Setting DRQ\n");
				buf.id_drq = buf1.id_drq;
				modified = TRUE;
			}
			if (buf.id_maddr != (caddr_t)-1 && buf.id_maddr != buf1.id_maddr) {
				if (verbose)
					printf("Setting memory addres\n");
				buf.id_maddr = buf1.id_maddr;
				modified = TRUE;
			}
			if (buf.id_msize != buf1.id_msize) {
				if (verbose)
					printf("Setting msize\n");
				buf.id_msize = buf1.id_msize;
				modified = TRUE;
			}
			if (buf.id_flags != buf1.id_flags) {
				if (verbose)
					printf("Setting flags\n");
				buf.id_flags = buf1.id_flags;
				modified = TRUE;
			}
			if (buf.id_enabled != buf1.id_enabled) {
				if (verbose)
					printf("Setting device enable/disable\n");
				buf.id_enabled = buf1.id_enabled;
				modified = TRUE;
			}
			if (verbose)
				printf("----------------------------------------------------\n");
			if (modified && !testonly) {

				res = lseek(f, -(off_t) sizeof(struct isa_device),
					    SEEK_CUR);
				if (write(f, &buf, sizeof(struct isa_device)) <= 0)
					fatal("write", NULL);

			}
		} while (buf.id_id != 0 && buf1.id_id != 0);
	}

	if (kvm_nlist(kd, nlaux) != 0) {
		/* num_eisa_conf need not exist, only handle it if found */
		if (verbose)
			printf("num_eisa_slots not found, ignoring.\n");
	} else {
		if (nlaux[0].n_type == 0)
			fatal("kvm_nlist", "bad symbol type");
		pos1 = nlaux[0].n_value;
		if (kvm_read(kd, pos1, &s, sizeof(int)) < 0)
			fatal("kvmread", NULL);

		if (nlist(kernel, nlaux) != 0)
			fatal("nlist", NULL);
		if (nlaux[0].n_type == 0)
			fatal("nlist", "bad symbol type");
		pos = nlaux[0].n_value + getpagesize() - entry;
		if (lseek(f, pos, SEEK_SET) != pos)
			fatal("seek", NULL);
		if ((res = read(f, (char *) &i, sizeof(int)))
		    <= 0)
			fatal("read", NULL);

		if (i != s) {
			if (verbose)
				printf("\nChanging num_eisa_slots from %d to %d.\n",
				       i, s);
			if (!testonly) {
				res = lseek(f, -(off_t) sizeof(int),
					    SEEK_CUR);
				if (write(f, &s, sizeof(int)) <= 0)
					fatal("write", NULL);
			}
		}
	}

	if (kvm_nlist(kd, nlpnp) != 0) {
		/* pnp_ldn_overrides need not exist, only handle it if found */
		if (verbose)
			printf("pnp_ldn_overrides not found, ignoring.\n");
	} else {
		if (nlpnp[0].n_type == 0)
			fatal("kvm_nlist", "bad symbol type");
		/* must handle it, read and write... */
		pos1 = nlpnp[0].n_value ;
		if (kvm_read(kd, pos1, &new_ov, sizeof(new_ov) ) < 0 )
			fatal("kvmread", NULL);
		if (nlist(kernel, nlpnp) != 0)
			fatal("nlist", NULL);
		if (nlpnp[0].n_type == 0)
			fatal("nlist", "bad symbol type");
		pos = nlpnp[0].n_value + getpagesize() - entry;
		if (lseek(f, pos, SEEK_SET) != pos)
			fatal("seek", NULL);
		if ((res = read(f, (char *) &old_ov, sizeof(old_ov))) <= 0 )
			fatal("read", NULL);
		if (testonly || bcmp(&old_ov, &new_ov, sizeof(old_ov)) ) {
		    if (verbose) {
			int i;
			printf("CSN LDN conf en  irq  drq vendor_id. ....\n");
			for (i=0; i < MAX_PNP_LDN ; i++) {
			    if (new_ov[i].csn>0 && new_ov[i].csn<255) {
				int maxp, maxm;
				for (maxp=7; maxp>=0 ; maxp--)
				    if (new_ov[i].port[maxp] !=0) break;
				for (maxm=3; maxm>=0 ; maxm--)
				    if (new_ov[i].mem[maxm].base !=0) break;
				printf("%3d %3d %-4s %c ",
				    new_ov[i].csn,
				    new_ov[i].ldn,
				    new_ov[i].override ? "OS":"BIOS",
				    new_ov[i].enable ? 'y':'n');
				printf(new_ov[i].irq[0] == 0 ? "  -":" %2d",
				   new_ov[i].irq[0]);
				printf(new_ov[i].irq[1] == 0 ? " - ":"%-2d",
				   new_ov[i].irq[1]);
				printf(new_ov[i].drq[0] == 4 ? " -":" %d",
				   new_ov[i].drq[0]);
				printf(new_ov[i].drq[1] == 4 ? " -":" %d",
				   new_ov[i].drq[1]);
				if (new_ov[i].vendor_id > 0)
				    printf(" 0x%08x", new_ov[i].vendor_id);
				if (new_ov[i].flags > 0)
				    printf(" flags 0x%08lx", new_ov[i].flags);
				if (maxp >=0) {
				    int j;
				    printf(" port 0x%x", new_ov[i].port[0]);
				    for(j=1;j<=maxp;j++)
					printf(" 0x%x", new_ov[i].port[j]);
				}
				if (maxm >=0) {
				    int j;
				    printf(" mem 0x%lx", new_ov[i].mem[0].base);
				    for(j=1;j<=maxm;j++)
					printf(" 0x%lx", new_ov[i].mem[j].base);
				}
				printf("\n");
			    }
			}
		    }
		    if (!testonly) {
			res = lseek(f, -(off_t) sizeof(new_ov),
					    SEEK_CUR);
			if (write(f,&new_ov, sizeof(new_ov)) <= 0)
					fatal("write", NULL);
		    }
		}
	}
	if (chflags(kernel, flags) < 0)
		fatal("chflags restore", NULL);

	kvm_close(kd);
	close(f);
	return(0);
}
